### PR TITLE
[NEW] Settings view

### LINKED ISSUE #803: Change "SettingsView" layout
We need to add more settings too, but this layout isn't good.

### SettingsView needs:
- Contact us
- Language
- Theme
- Share this app
- License
- Version (it's on SidebarView now and it should be here)
- Server version
- Send crash report

### PR DESCRIPTION
**Pull request by pranavpandey1998official**
@RocketChat/ReactNative

WhatsApp Image 2019-05-16 at 3 54 08 PM

ezgif com-video-to-gif (6)
Closes #803

### CODE REVIEW

File: app/sagas/login.js [hunk 1]
```diff
@@ -85,8 +85,18 @@ const handleLogout = function* handleLogout() {
 	}
 };
 
-const handleSetUser = function handleSetUser({ user }) {
+const handleSetUser = function* handleSetUser({ user }) {
```
Comment (diegolmello) [inline app/sagas/login.js]: Why?

File: app/views/SettingsView/index.js [hunk 2]
```diff
@@ -1,184 +1,179 @@
...
-			setTimeout(() => {
-				showErrorAlert(I18n.t('There_was_an_error_while_action', { action: I18n.t('saving_preferences') }));
-				log('saveUserPreferences', e);
-			}, 300);
-		}
-	}
+	renderLastSection = ({ item }) => (
+		<View style={[styles.sectionItem, item.disable && styles.sectionItemDisabled]}>
+			<View style={{ flex: 1 }}>
+				<Text style={{ ...styles.sectionItemTitle, marginTop: 5 }}>{item.title}</Text>
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/views/SettingsView/index.js]: Also, never destructure styles like this. There's no need.
Style already has support to arrays.
It's wrong, but you should have done this way:
```jsx
<Text style={[styles.sectionItemTitle, { marginTop: 5 }]}>{item.title}</Text>
```

File: app/sagas/login.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (pranavpandey1998official) [reply app/sagas/login.js]: K.. so the one of problem that @kareemhamdy100  was facing is that when first time you open app after changing the language it is showing old language but then after you open it again it will show correct language soo problem lies https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/944e4198bbaf95e78b2be6b6a1270029f0fa226a/app/sagas/selectServer.js#L40 see it is setting user settings from local storage the whole login and changing of internal setting happens after that (this internal storage is changed only on login ) so I added condition to change this on change of user setting

**Comment by pranavpandey1998official** [comment]
WhatsApp Image 2019-05-22 at 2 06 56 AM (1)
WhatsApp Image 2019-05-22 at 2 06 56 AM

File: app/views/SettingsView/styles.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import { StyleSheet } from 'react-native';
+
+import {
+	COLOR_SEPARATOR, COLOR_WHITE, COLOR_BORDER
+} from '../../constants/colors';
+import sharedStyles from '../Styles';
+
+export default StyleSheet.create({
```
Comment (diegolmello) [inline app/views/SettingsView/styles.js]: Do it like this:



Look at the borders, item height, margins, backgroundColor.
Also, remove crash sections for now.

File: app/views/SettingsView/index.js [hunk 4]
```diff
@@ -1,238 +1,163 @@
...
+
+					{this.renderItem({ title: I18n.t('Contact_us'), onPress: this.sendEmail, showActionIndicator: true })}
+					{this.renderSeparator()}
+					{this.renderItem({ title: I18n.t('Language'), onPress: this.navigateToRoom('LanguageView'), showActionIndicator: true })}
+					{this.renderSeparator()}
+					{this.renderItem({ title: I18n.t('Theme'), showActionIndicator: true, disable: true })}
+					{this.renderSeparator()}
+					{this.renderItem({ title: I18n.t('Share_this_app'), showActionIndicator: true, disable: true })}
+					{this.renderSeparator()}
+					{this.renderItem({ title: I18n.t('Theme'), showActionIndicator: true, disable: true })}
```
Comment (diegolmello) [inline app/views/SettingsView/index.js]: Why did you add two themes?

File: app/sagas/login.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (pranavpandey1998official) [reply app/sagas/login.js]: @diegolmello without it Language change will still have bugs ..

[the remark this answers is not in the document]
Comment (pranavpandey1998official) [reply app/sagas/login.js]: I am opening a issue for this

File: app/views/SettingsView/index.js [hunk 5]
```diff
@@ -1,238 +1,163 @@
...
-				keyboardVerticalOffset={128}
-			>
+			<SafeAreaView style={styles.container} forceInset={{ bottom: 'never' }}>
 				<StatusBar />
 				<ScrollView
-					contentContainerStyle={sharedStyles.containerScrollView}
-					testID='settings-view-list'
 					{...scrollPersistTaps}
+					contentContainerStyle={styles.contentContainer}
+					showsVerticalScrollIndicator={false}
```
[the remark this answers is not in the document]
Comment (pranavpandey1998official) [reply app/views/SettingsView/index.js]: the indicator wasn't looking good

File: app/views/LanguageView/index.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (pranavpandey1998official) [reply app/views/LanguageView/index.js]: done

File: app/views/SettingsView/index.js [hunk 7]
```diff
@@ -1,238 +1,163 @@
...
-			return;
-		}
-
-		const params = {};
+	renderSeparator = () => <View style={styles.separator} />;
 
-		// language
-		if (userLanguage !== language) {
-			params.language = language;
+	renderItem = (item) => {
```
[the remark this answers is not in the document]
Comment (pranavpandey1998official) [reply app/views/SettingsView/index.js]: done

File: app/views/SettingsView/index.js [hunk 8]
```diff
@@ -1,238 +1,163 @@
...
+		);
 	}
 
-	toggleMarkdown = (value) => {
-		AsyncStorage.setItem(MARKDOWN_KEY, JSON.stringify(value));
-		const { toggleMarkdown } = this.props;
-		toggleMarkdown(value);
-		Answers.logCustom('toggle_markdown', { value });
-	}
+	renderItemSwitch = item => (
```
[the remark this answers is not in the document]
Comment (pranavpandey1998official) [reply app/views/SettingsView/index.js]: done

File: app/views/SettingsView/styles.js (hunk 3 above)
Comment (pranavpandey1998official) [reply app/views/SettingsView/styles.js]: done

File: app/views/SettingsView/index.js (hunk 4 above)
Comment (pranavpandey1998official) [reply app/views/SettingsView/index.js]: removed

File: app/views/SettingsView/index.js [hunk 9]
```diff
@@ -1,238 +1,163 @@
...
 
-		// language
-		if (userLanguage !== language) {
-			params.language = language;
+	renderItem = (item) => {
+		if (!item.onPress) {
+			item.onPress = () => {};
 		}
+		return (
+			<RectButton
```
[the remark this answers is not in the document]
Comment (pranavpandey1998official) [reply app/views/SettingsView/index.js]: done

**Comment by pranavpandey1998official** [comment]
Simulator Screen Shot - iPhone X - 2019-05-28 at 18 11 42
Simulator Screen Shot - iPhone X - 2019-05-28 at 18 15 00

File: app/views/LanguageView/index.js [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+import Loading from '../../containers/Loading';
+import { showErrorAlert, Toast } from '../../utils/info';
+import log from '../../utils/log';
+import { setUser as setUserAction } from '../../actions/login';
+import StatusBar from '../../containers/StatusBar';
+import { CustomIcon } from '../../lib/Icons';
+import styles from './styles';
+import { COLOR_TEXT } from '../../constants/colors';
+
+const LANGUAGES = [{
```
Comment (diegolmello) [inline app/views/LanguageView/index.js]: Order alphabetically.

File: app/views/LanguageView/index.js [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+						<Text style={styles.text}>{label}</Text>
+						{isSelected ? <CustomIcon name='check' size={20} style={styles.checkIcon} /> : null }
+					</View>
+				</RectButton>
+				{this.renderSeparator()}
+			</React.Fragment>
+		);
+	}
+
+	render() {
```
Comment (diegolmello) [inline app/views/LanguageView/index.js]: It's missing a grey background.
Only items should be white.

File: app/views/LanguageView/index.js [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+				keyboardVerticalOffset={128}
+			>
+				<StatusBar />
+				<ScrollView
+					contentContainerStyle={styles.containerScrollView}
+					testID='settings-view-list'
+					{...scrollPersistTaps}
+				>
+					<SafeAreaView style={sharedStyles.container} testID='settings-view' forceInset={{ bottom: 'never' }}>
+						{LANGUAGES.map(item => this.renderItem(item))}
```
Comment (diegolmello) [inline app/views/LanguageView/index.js]: FlatList

Comment (diegolmello) [inline app/views/LanguageView/index.js]: First item must have a top separator without left margin
Last item must have a bottom separator without left margin

File: app/views/LanguageView/styles.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+		flex: 1,
+		padding: 18,
+		marginHorizontal: 4,
+		flexDirection: 'row',
+		alignItems: 'center',
+		justifyContent: 'space-between'
+	},
+	text: {
+		fontSize: 16,
+		color: COLOR_TEXT
```
Comment (diegolmello) [inline app/views/LanguageView/styles.js]: Inconsistent text style.
You must use `textRegular` and `textColorNormal`

File: app/views/SettingsView/ButtonWithSwitch.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
+import { View, Text, Switch } from 'react-native';
+import PropTypes from 'prop-types';
+
+import styles from './styles';
+import { COLOR_DANGER, COLOR_SUCCESS } from '../../constants/colors';
+import { isAndroid } from '../../utils/deviceInfo';
+
+const ButtonWithSwitch = React.memo(({
+	onValueChange, title, subTitle, disable, value
```
Comment (diegolmello) [inline app/views/SettingsView/ButtonWithSwitch.js]: subtitle

File: app/views/SettingsView/index.js [hunk 15]
```diff
@@ -1,236 +1,108 @@
...
-		}
-		return null;
+	sendEmail = () => {
+		const subject = encodeURI('React Native App Support');
+		const email = encodeURI('support@rocket.chat');
+		const description = encodeURI(`
+			version: ${ getReadableVersion }
+			device: ${ getDeviceModel }
+		`);
+		Linking.openURL(`mailto:${ email }?subject=${ subject }&body=${ description }`);
```
Comment (diegolmello) [inline app/views/SettingsView/index.js]: `try/catch` with an alert.

File: app/views/SettingsView/index.js [hunk 16]
```diff
@@ -1,236 +1,108 @@
...
 	}
 
-	formIsChanged = () => {
-		const { userLanguage } = this.props;
-		const { language } = this.state;
-		return !(userLanguage === language);
-	}
+	openLink = link => () => openLink(link)
```
Comment (diegolmello) [inline app/views/SettingsView/index.js]: Why not `openLink = link => openLink(link)`?

File: app/views/SettingsView/index.js [hunk 17]
```diff
@@ -1,236 +1,108 @@
...
-			}, 300);
-		} catch (e) {
-			this.setState({ saving: false });
-			setTimeout(() => {
-				showErrorAlert(I18n.t('There_was_an_error_while_action', { action: I18n.t('saving_preferences') }));
-				log('err_save_user_preferences', e);
-			}, 300);
-		}
-	}
+	renderSectionSeparator = () => <View style={styles.sectionSeparatorBorder} />;
```
Comment (diegolmello) [inline app/views/SettingsView/index.js]: Can you create a global component using `React.memo`?

File: app/views/SettingsView/styles.js [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+		opacity: 0.3
+	},
+	separator: {
+		height: StyleSheet.hairlineWidth,
+		backgroundColor: COLOR_SEPARATOR,
+		marginHorizontal: 10
+	},
+	sectionSeparatorBorder: {
+		borderColor: '#F6F7F9',
+		borderWidth: 5
```
Comment (diegolmello) [inline app/views/SettingsView/styles.js]: Bad practice.
This way you're rendering a view with empty height and border width 5.

It should be only a view with height and background.

File: app/views/SettingsView/styles.js [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+	sectionItemDisabled: {
+		opacity: 0.3
+	},
+	separator: {
+		height: StyleSheet.hairlineWidth,
+		backgroundColor: COLOR_SEPARATOR,
+		marginHorizontal: 10
+	},
+	sectionSeparatorBorder: {
+		borderColor: '#F6F7F9',
```
Comment (diegolmello) [inline app/views/SettingsView/styles.js]: Inline color

File: app/views/SettingsView/index.js [hunk 20]
```diff
@@ -1,236 +1,108 @@
...
+
+					<Button title={I18n.t('License')} onPress={this.openLink(LICENSE_LINK)} showActionIndicator />
+					{this.renderSeparator()}
+					<Button title={I18n.t('Version_no', { version: getReadableVersion })} />
+					{this.renderSeparator()}
+					<Button title={I18n.t('Server_version', { version: server.version })} subTitle={`${ server.server.split('//')[1] }`} />
+
+					{this.renderSectionSeparator()}
+
+					<ButtonWithSwitch title={I18n.t('Enable_markdown')} value={useMarkdown} onValueChange={this.toggleMarkdown} />
```
Comment (diegolmello) [inline app/views/SettingsView/index.js]: Margin right

File: app/views/SettingsView/index.js [hunk 21]
```diff
@@ -1,236 +1,108 @@
...
+					{this.renderSeparator()}
+					<Button title={I18n.t('Crash_report_disclaimer')} disable />
```
Comment (diegolmello) [inline app/views/SettingsView/index.js]: Why is this a button if it's not touchable?
Even if it's disabled, it shouldn't be a button.
You should use semantics here.

Natively the device will build a button and apply a disable prop.
It should be a simple view.

File: app/views/SettingsView/index.js [hunk 22]
```diff
@@ -1,236 +1,108 @@
...
+					<Button title={I18n.t('License')} onPress={this.openLink(LICENSE_LINK)} showActionIndicator />
+					{this.renderSeparator()}
+					<Button title={I18n.t('Version_no', { version: getReadableVersion })} />
+					{this.renderSeparator()}
+					<Button title={I18n.t('Server_version', { version: server.version })} subTitle={`${ server.server.split('//')[1] }`} />
+
+					{this.renderSectionSeparator()}
+
+					<ButtonWithSwitch title={I18n.t('Enable_markdown')} value={useMarkdown} onValueChange={this.toggleMarkdown} />
+					{this.renderSeparator()}
```
Comment (diegolmello) [inline app/views/SettingsView/index.js]: I added Firebase in a previous PR, so you can enable that toggle crash switch.
https://rnfirebase.io/docs/v5.x.x/crashlytics/reference/crashlytics#enableCrashlyticsCollection

File: app/views/SettingsView/index.js (hunk 17 above)
Comment (pranavpandey1998official) [reply app/views/SettingsView/index.js]: for this View or for every View?

File: app/views/SettingsView/index.js [hunk 23]
```diff
@@ -1,236 +1,120 @@
...
-	}
+	renderSeparator = () => <View style={styles.separator} />;
 
-	toggleMarkdown = (value) => {
-		AsyncStorage.setItem(MARKDOWN_KEY, JSON.stringify(value));
-		const { toggleMarkdown } = this.props;
-		toggleMarkdown(value);
-		firebase.analytics().logEvent('toggle_markdown', { value });
-	}
+	renderCrashReportDisclaimer = () => (
```
Comment (diegolmello) [inline app/views/SettingsView/index.js]: Remove it.

File: app/views/SettingsView/index.js [hunk 24]
```diff
@@ -1,236 +1,120 @@
...
+					{this.renderSeparator()}
+					<Button title={I18n.t('Version_no', { version: getReadableVersion })} />
+					{this.renderSeparator()}
+					<Button title={I18n.t('Server_version', { version: server.version })} subTitle={`${ server.server.split('//')[1] }`} />
+
+					{this.renderSectionSeparator()}
+
+					<ButtonWithSwitch title={I18n.t('Enable_markdown')} value={useMarkdown} onValueChange={this.toggleMarkdown} />
+					{this.renderSeparator()}
+					{this.renderCrashReportDisclaimer()}
```
Comment (diegolmello) [inline app/views/SettingsView/index.js]: Remove it

File: app/views/SettingsView/index.js (hunk 20 above)
Comment (diegolmello) [reply app/views/SettingsView/index.js]: Why did you mark as resolved if you didn't?

File: app/views/SettingsView/index.js (hunk 23 above)
Comment (pranavpandey1998official) [reply app/views/SettingsView/index.js]: I'll be commenting this as it will be needed in the future

File: app/views/SettingsView/index.js (hunk 20 above)
Comment (pranavpandey1998official) [reply app/views/SettingsView/index.js]: I did resolve it https://github.com/pranavpandey1998official/Rocket.Chat.ReactNative/blob/e599ac7b781097b3b08f90ba63431ec571b4318b/app/views/SettingsView/styles.js#L49

File: e2e/14-setting.spec.js [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+		after(async() => {
+			takeScreenshot();
+		});
+	});
+
+	describe('Language', async() => {
+		it('should navigate to language view', async() => {
+			await element(by.id('settings-view-language')).tap();
+			await waitFor(element(by.id('language-view'))).toBeVisible().withTimeout(60000);
+            await expect(element(by.id('language-view-zh-CN'))).toExist();
```
Comment (diegolmello) [inline e2e/14-setting.spec.js]: Indentation

File: e2e/14-setting.spec.js [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+const { logout, navigateToLogin, login } = require('./helpers/app');
+
+describe('Settings screen', () => {
+	before(async() => {
+		await device.reloadReactNative();
+		await expect(element(by.id('rooms-list-view'))).toBeVisible();
+		await element(by.id('rooms-list-view-sidebar')).tap();
+		await waitFor(element(by.id('sidebar-view'))).toBeVisible().withTimeout(2000);
+		await waitFor(element(by.id('sidebar-settings'))).toBeVisible().withTimeout(2000);
+		// await expect(element(by.id('sidebar-profile'))).toBeVisible();
```
Comment (diegolmello) [inline e2e/14-setting.spec.js]: Remove
